Unused trait imports (formerly anonymous trait import)

RuairidhWilliamson opened the pull request:
For #11969

I'm looking for help and feedback on implementing a new lint for suggesting `use ... as _` for traits where possible.

I have had a go at implementing this but I don't know if this is the best way to do it as I am new to clippy.

There are some edge cases I can think of where this doesn't work but have aired on the side of false negatives instead of false positives.

An example of a false negative. I couldn't figure out the best way to resolve an import from within clippy. The sub module imports MyAny so that cannot be anonymized but `use std::any::Any` could be. In this case it is not caught because `Any` and `MyAny` have the same DefId.
```rust
mod nested_mod_used_bad1 {
    use std::any::Any;
    use std::any::Any as MyAny;
    mod foo {
        use crate::nested_mod_used_bad1::MyAny;
        fn foo() {
            println!("{:?}", MyAny::type_id("foo"));
        }
    }
}
```

Any feedback is much appreciated.

-------
changelog: new lint: `unused_trait_names`

y21 commented:
I've not properly looked at the changes here in full detail but some initial comments:

> There are some edge cases I can think of where this doesn't work but have aired on the side of false negatives instead of false positives.

I'd say that this is fine and even preferred (false negatives over false positives), especially for a lint that's going to be as trigger-heavy as this one.


I wonder though, instead of reimplementing a part of resolution, could this use `tcx.maybe_unused_trait_imports().contains(use_def_id)`? If I remember right, that set is created in early name resolution for imports that aren't referred to by path but may be used for method resolution in typeck, which sounds like it might be of use here?
The current way of walking the whole module and all its contents for every single import seems like it could get expensive, but that might just be premature

RuairidhWilliamson commented:
Thank you for the response.

You are right. `maybe_unused_trait_imports` solves exactly this problem and I was able to remove the import resolution. I am not sure how I missed that query but it has made the code much cleaner.

I had to change some of the tests but I think they are correct now and fixed the false negatives.

nyurik commented:
Awesome lint, thanks!!!  I would even suggest removing the `wip` from the subject.

You used `Applicability::MaybeIncorrect` - are there any such known bad false positive cases?  If so, you may want to add them, at least as a disabled code block, to the tests - so that in the future they can be handled.  Also, do add all the false negatives as tests - even though they will be ignored for now, others may improve their handling.

Lastly, a tiny minor nit - you may want to move the `main() {}` to the top of the test file - this way new test cases will be added at the end, and the line numbering will stay intact in the blessed results.

RuairidhWilliamson commented:
Thanks @nyurik :)

Yes I agree it shouldn't be wip anymore.

> You used `Applicability::MaybeIncorrect` - are there any such known bad false positive cases? If so, you may want to add them, at least as a disabled code block, to the tests - so that in the future they can be handled. Also, do add all the false negatives as tests - even though they will be ignored for now, others may improve their handling.

Yes, good point. The change to using `maybe_unused_trait_imports()` has cleared up the edge cases I can think of so I will make it `MachineApplicable`.

> Lastly, a tiny minor nit - you may want to move the `main() {}` to the top of the test file - this way new test cases will be added at the end, and the line numbering will stay intact in the blessed results.

Ah that is a good point. I have moved main to the top.

y21 commented on clippy_lints/src/anon_trait_import.rs:
````diff
@@ -0,0 +1,72 @@
+    ///
+    /// ### Example
+    /// ```no_run
+    ///   use std::fmt::Write;
````
Can you remove the extra indentation from the examples here? The rendered markdown currently looks like this
![image](https://github.com/user-attachments/assets/7c0099a9-1fa1-46b9-b1cf-d5311fe087cb)

y21 reviewed (commented):
A lot of the comments that I had before when I looked at this first are no longer relevant and fixed now with the extra visitor removed so that's nice :)

y21 commented:
> You used `Applicability::MaybeIncorrect` - are there any such known bad false positive cases?

FWIW, something like `use path::to::Trait::{self};` currently has a wrong/breaking suggestion (suggests `use path::to::Trait as _::{self}`), but it seems like a weird thing to write, considering that trait-relative items can't be imported, so `{self}` would always be on its own. So I'd say it's fine to ignore that and not block machine-applicability on it

RuairidhWilliamson commented on clippy_lints/src/anon_trait_import.rs:
```diff
@@ -0,0 +1,72 @@
+            && let module = cx.tcx.parent_module_from_def_id(item.owner_id.def_id)
+            && cx.tcx.visibility(item.owner_id.def_id) == Visibility::Restricted(module.to_def_id())
+            && let Some(last_segment) = path.segments.last()
+            && let Some(snip) = snippet_opt(cx, last_segment.ident.span)
```
For some reason using `is_from_proc_macro(cx, item)` wouldn't trigger when testing with `with_span!`. I managed to get it working by using `is_from_proc_macro(cx, &last_segment.ident)` but I don't understand why that worked and the former didn't.

RuairidhWilliamson commented:
Thanks for the review.

> > You used `Applicability::MaybeIncorrect` - are there any such known bad false positive cases?
> 
> FWIW, something like `use path::to::Trait::{self};` currently has a wrong/breaking suggestion (suggests `use path::to::Trait as _::{self}`), but it seems like a weird thing to write, considering that trait-relative items can't be imported, so `{self}` would always be on its own. So I'd say it's fine to ignore that and not block machine-applicability on it

Ah yes that is an interesting edge case. I have had a look at the HIR to see if it is easy to ignore this pattern but it seems to be hard to distinguish at that level.
```
use std::any::Any::{self};
```
```
use ::{};
use std::any::Any;
```

I guess maybe I could use the span to check the source manually but that feels messy.

If you are happy that it is not an issue that needs to be addressed then I am too :)

bors commented:
:umbrella: The latest upstream changes (presumably #13247) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #12987) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

y21 commented on clippy_lints/src/anon_trait_import.rs:
````diff
@@ -0,0 +1,94 @@
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub ANON_TRAIT_IMPORT,
````
```suggestion
    pub ANON_TRAIT_IMPORTS,
```
Lint names should generally be in plural so it makes sense when read as "allow anon trait imports"

y21 commented on clippy_lints/src/anon_trait_import.rs:
```diff
@@ -0,0 +1,94 @@
+    /// Checks for `use Trait` where the Trait is only used for its methods and not referenced by a path directly.
+    ///
+    /// ### Why is this bad?
+    /// Traits imported that aren't used directly can be imported anonymously `use Trait as _`.
+    /// It is more explicit, avoids polutting the current scope with unused names and can be useful to show which imports are required for traits.
```
```suggestion
    /// Traits imported that aren't used directly can be imported anonymously with `use Trait as _`.
    /// It is more explicit, avoids polluting the current scope with unused names and can be useful to show which imports are required for traits.
```

y21 commented on clippy_lints/src/anon_trait_import.rs:
````diff
@@ -0,0 +1,94 @@
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub ANON_TRAIT_IMPORT,
+    nursery,
````
Do you intend to move this to a different category before the merge or keep it in the nursery for now?

Usually new lints start out in a more defined category. Nursery is typically for lints that are known to be buggy

RuairidhWilliamson replied:
Good point, I have moved it to pedantic.

y21 reviewed (approved):
Looks good to me aside from a few small things

y21 commented:
This might be more or less tangential to this lint but we might also want to adjust rustc's suggestion for importing traits for methods. I imagine it would be somewhat annoying to apply the compiler's suggestion to import a trait only to immediately trigger a clippy lint.

nyurik commented:
@y21 i agree, `rustc` should suggest `use Foo as _;` rather than `use Foo` when the trait method is being used. This would both prevent inconsistencies with this lint, as well as teach users about namespace pollution.  Would you like to create an issue for that?

RuairidhWilliamson commented:
I have found another sub optimal scenario but don't think it is bad enough to prevent merge.

It would be nice if multiple renames that import the same trait gets suggested to combine into a single import. Currently this produces a suggestion for each import meaning running `--fix` produces the second output.

```rust
use std::any::{Any, Any as MyAny};
```
After running `--fix`:
```rust
use std::any::{Any as _, Any as _};
```
Rustc will lint one of the imports as unused so should highlight the issue.

bors commented:
:umbrella: The latest upstream changes (presumably #13369) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

y21 commented:
> It would be nice if multiple renames that import the same trait gets suggested to combine into a single import. Currently this produces a suggestion for each import meaning running --fix produces the second output.

I think that's fine. Or at least, I don't think this situation would be common enough to be an actual problem. It seems odd to import the same trait twice to begin with IMO.

y21 commented on tests/ui/anon_trait_imports.rs:
Please also add a test for making sure that `#[allow(clippy::anon_trait_imports)]` on the use item works. You might have to include `anon_trait_imports` here to avoid a false positive on the useless_attribute lint: https://github.com/rust-lang/rust-clippy/blob/131681b4b9c2460d223b5939a273f0356a8e06f5/clippy_lints/src/attrs/useless_attribute.rs#L44-L46

RuairidhWilliamson replied:
added :+1:

RuairidhWilliamson replied:
Good point, added.

There is a false positive for `useless_attribute` in this case

y21 replied:
Being able to allow the lint on the item that it warns on is something that should really work for all lints (without triggering other lints), so it would be good if you could fix that false positive for this lint. It should be fairly straightforward, you just need to add it to the `matches!` macro (see initial comment)

RuairidhWilliamson replied:
Sorry, yeah I forgot that you explained how to fix that. Fixed :+1:

y21 commented:
Opened a thread on zulip for the FCP process that new lints need to go through: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/FCP.3A.20anon_trait_imports

Implementation looks good to me aside from the test change that I commented on, but that shouldn't involve any major changes in a way that would block the FCP.

kpreid commented:
Isn't this lint named backwards? Lints are named for what they report, not what they suggest you do instead, and this lint is objecting to unused names and suggesting an anonymous import.

How about naming it `unused_trait_names`?

bors commented:
:umbrella: The latest upstream changes (presumably #13313) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

RuairidhWilliamson commented:
Based on feedback in zulip I've renamed the lint from `anon_trait_imports` -> `unused_trait_names` and moved the lint from `pedantic` to `restriction` (which means it doesn't get run as dogfood).

nyurik commented:
we could probably add it to `Cargo.toml` to be part of dogfood? I really think this is a good thing :tm:  to have ...  so if we make this more visible, and if we make the compiler suggest code in this format, eventually it could make it to pedantic and then to style...

RuairidhWilliamson commented:
> we could probably add it to `Cargo.toml` to be part of dogfood? I really think this is a good thing ™️ to have ... so if we make this more visible, and if we make the compiler suggest code in this format, eventually it could make it to pedantic and then to style...

yeah I agree but maybe it is best to do it in a follow up PR as it adds a lot of changes across clippy :)

y21 commented on clippy_lints/src/unused_trait_names.rs:
```diff
@@ -0,0 +1,94 @@
+}
+
+pub struct UnusedTraitNames {
+    msrv: Msrv,
```
The lint also needs to be added to the list here now that it respects the MSRV configuration (forgot to mention that earlier) https://github.com/rust-lang/rust-clippy/blob/9be28b143993f48cd0756778b28244763cd72e48/clippy_config/src/conf.rs#L510

y21 commented on clippy_lints/src/unused_trait_names.rs:
````diff
@@ -0,0 +1,94 @@
+    ///     println!("{s}");
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
````
```suggestion
    #[clippy::version = "1.83.0"]
```

y21 reviewed (approved):
Looks good to me now aside from two small things. Can you also squash the commits here a bit?

I also agree that if we're going to apply this style to the code base and enable the lint it should be done in a separate PR since it'll probably involve a discussion, which doesn't really need to block the lint itself

RuairidhWilliamson commented:
I've addressed your review and squashed everything into two commits :)

y21 commented:
Great, thank you for implementing this lint!

@bors r+

bors commented:
:pushpin: Commit 739ef7bf0d5686ef10e7700f34e34995936e00ff has been approved by `y21`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=y21 739ef7bf0d5686ef10e7700f34e34995936e00ff -->
<!-- homu: {"type":"Approved","sha":"739ef7bf0d5686ef10e7700f34e34995936e00ff","approver":"y21","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 739ef7bf0d5686ef10e7700f34e34995936e00ff with merge 7951e02a6a0f3d0eb2b18690f8ab6556662caa0b...
<!-- homu: {"type":"BuildStarted","head_sha":"739ef7bf0d5686ef10e7700f34e34995936e00ff","merge_sha":"7951e02a6a0f3d0eb2b18690f8ab6556662caa0b"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/10981883286/job/30489484000)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/actions/runs/10981883286/job/30489484000","builder_name":"checks-action_test"} -->

y21 commented:
PR description was missing a `changelog:` line, I edited it in.

@bors retry

bors commented:
:hourglass: Testing commit 739ef7bf0d5686ef10e7700f34e34995936e00ff with merge abdf173ef2972033e1b5ac78e1bbc886b7f871c1...
<!-- homu: {"type":"BuildStarted","head_sha":"739ef7bf0d5686ef10e7700f34e34995936e00ff","merge_sha":"abdf173ef2972033e1b5ac78e1bbc886b7f871c1"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/10981908079/job/30489545856), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/10981908081/job/30489549054), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/10981908077/job/30489631485)
Approved by: y21
Pushing abdf173ef2972033e1b5ac78e1bbc886b7f871c1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"y21","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/10981908077/job/30489631485","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/10981908079/job/30489545856","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/10981908081/job/30489549054"},"merge_sha":"abdf173ef2972033e1b5ac78e1bbc886b7f871c1"} -->

RuairidhWilliamson commented:
@y21 Great thank you for all your help, I'm glad we got this added!
